flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java

```diff
@@ -343,13 +396,30 @@ public Builder field(String name, TypeInformation<?> typeInfo) {
 			return field(name, fromLegacyInfoToDataType(typeInfo));
 		}
 
+		/**
+		 * Specifies the previously defined field as an event-time attribute and specifies the watermark strategy.
+		 *
+		 * @param rowtimeAttribute the field name as a rowtime attribute, can be a nested field using dot separator.
+		 * @param watermarkStrategy the watermark strategy expression
+		 */
+		public Builder watermark(String rowtimeAttribute, ResolvedExpression watermarkStrategy) {
```

Comment:
If user using this `Builder` to manually specify watermark, how can he get `ResolvedExpression`. AFAIK if he uses scala dsl, it can only generate `UnresolvedCallExpression` or `UnresolvedReferenceExpression`

Reply:
Yes. You are right. There are 2 reasons here:
1) we can't get `ExpressionResolver` in builder to resolve expression.
2) `TableSchema.Builder` is going to be internally used only.  As we are planning to remove {{TableSource#getTableSchema}}, users should don't need to create `TableSchema` anymore.

Reply:
I added an `@Internal` annotation on this method.

Reply:
So besides from DDL way, how to user specify watermark programmatically?

Reply:
Using the `Schema` descriptor API.

Reply:
We will provide some API to `Schema` to define watermark, for example, `Schema.watermark("ts", 'ts - 5.seconds)`, then the planner will convert the unresolved expression to resolved expression using the function catalog, and then create the `TableSchema` by planner.

Reply:
It doesn't feel right when user just creating a schema should involves table planner..

Reply:
`Schema` is a pure API class which just defines metadata information and will not involve planner. When planner will be involved when we convert the `Schema` descriptor to `CatalogTable`, this is similar to convert DDL to `CatalogTable`.

Reply:
The `Schema` might be an easier way to construct the `TableSchema` but can we forbidden user from creating `TableSchema` by himself?

Reply:
If all the Flink API doesn't require `TableSchema` as parameter/input in the future, then I think users don't have a reason to create it. 

In FLIP-70, we will add computed column information into `TableSchema` too. The computed  expression should also be a resovled expression, because we need to know the datatype of this field.

Reply:
The problem is there are two major types of `Expression` now, one is unresolved and another one is resolved. We will only have interface (like Java DSL style) for unresolved expression. But the issue is we need some other information which only accessible from resolved one, like data types.

Reply:
As we discussed in FLIP-80, we will use the serialize (resolved) expression to string, this happens when serialize CatalogTable to properties. And deserialize string to unresolved expression, this happens when create `CatalogTable` from properties in `Catalog`. However, we missed that converting unresovled expression to resolved expression requires `ExpressionResolver` which is not held by `Catalog`. If the TableSchema builder requires `ResolvedExpression`, it's impossible to create a `TableSchema` or `CatalogTable` for `Catalog`. 

Let me sum up, I agree with @KurtYoung that using `ResolveExpression` here may can't work in the future, so I would like change it to :

```
public Builder watermark(String rowtimeAttribute, String watermarkExpression, DataType watermarkExprOutputType) {
```

The `ResolvedExpression getWatermarkStrategy()` in `WatermarkSpec` will be removed, and will expose `String getWatermarkExpressionString()` and `DataType getWatermarkExprOutputType()`.

What do you think?

Reply:
I think using plain strings in Java classes should be avoided in general. This is not very object oriented. I understand that constructing `ResolvedExpression` might be difficult at certain locations but we could also use just `Expression`. Actually, a SQL string is also just an unresolved expression until it has been parsed and validated.

Reply:
I'm fine to use `Expression` here and introduce an unresolved `SqlExpression` to wrap it. 
So does the following signature looks good to you? 

```java
public Builder watermark(String rowtimeAttribute, Expression watermarkExpression, DataType watermarkExprOutputType) {
```